### PR TITLE
VP9 SVC layers mapping

### PR DESCRIPTION
**Pull request by logist322**
This PR implements solution discussed in the issue #404.

### CODE REVIEW

File: src/packet/vp9.rs [hunk 1]
```diff
@@ -13,6 +13,83 @@ const MAX_VP9REF_PICS: usize = 3;
...
+    ///         if let (true, Some(end_of_target_layer)) = (
+    ///             // Check whether we need this `temporal layer`.
+    ///             tid <= target_temporal,
+    ///             // Check whether we have information about the right
+    ///             // border of data for the target `spatial layer`.
+    ///             vp9_extra.layers_scheme[target_spatial as usize],
+    ///         ) {
+    ///             // Cut off unwanted data of `spatial layer`s with spatial
+    ///             // indeces higher than `target_spatial`.
+    ///             media_data.data.drain(end_of_target_layer..);
```
Comment (davibe) [inline src/packet/vp9.rs:69]: This is a great example!

If someone truly drains the data out he should also update the layer_scheme accordingly ?

**Comment by logist322** [comment]
@algesten Done!
Also created a new issue #411 after testing this task

File: src/packet/vp9.rs [hunk 2]
```diff
@@ -13,6 +13,83 @@ const MAX_VP9REF_PICS: usize = 3;
...
+    pub tid: Option<u8>,
+
+    /// Map of the SVC layers widths.
+    ///
+    /// Specified for every spatial layer.
+    pub layers_widths: [Option<u16>; 3],
+
+    /// Map of the SVC layers heights.
+    ///
+    /// Specified for every spatial layer.
```
Comment (davibe) [inline src/packet/vp9.rs:86]: I wonder if it would be nicer to group `layer_*` into a struct by layer

File: src/packet/vp9.rs (hunk 1 above)
Comment (logist322) [reply src/packet/vp9.rs:69]: As far as changing `layers_scheme` does not have any affect on any mechanism and nothing relays on this `layers_scheme` in **str0m** (at least now), we can do anything with it after `MediaData` emitted.
In general, `layers_scheme` field has been added only to let **str0m** clients be able to select SVC layers they need. So, they should decide by themselves what to do with `MediaData::data` and `layers_scheme`.

Comment (logist322) [reply src/packet/vp9.rs:69]: I got you. You mean we should implement some synchronization between `MediaData::data` and `layers_scheme`, right?

File: src/packet/vp9.rs (hunk 2 above)
Comment (logist322) [reply src/packet/vp9.rs:86]: Or to join all information specific for the layer under the one `struct LayerInfo {position: usize, width: u16, height: u16}`?

Comment (davibe) [reply src/packet/vp9.rs:86]: Yes, I meant joining them. I am trying to imagine how it gets used

```
struct SpatialLayerInfo {
    start: usize, end: usize, width: u16, height: u16
}
```

One can use it like

```
let info = media_data.extra.layers.find(/* by resolution */);
let data_slice = media_data.data[info.start..info.end];
_ = decode(data_slice);
```

or

```
// remove layer 3
let info = media_data.extra.layers[3];
media_data.data.drain(info.start..);
media_data.extra.layers.drain(2..);
```

Comment (logist322) [reply src/packet/vp9.rs:86]: Yes, I got your idea. I tried to do the same at the beginning. But there is a problem with huge `MediaData` size. We can not add start field. In my current configuration it is already 440 bytes size.

Comment (davibe) [reply src/packet/vp9.rs:86]: I see

**Review by davibe — APPROVED** [review]

**Comment by algesten** [comment]
Perfect! Thanks!

**Pull request merged**